File: SimpleDownloadManager/DownloadManager.m
//
//  DownloadManager.m
//  SimpleDownloadManager
//
//  Created by Sam on 11/13/16.
//  Copyright © 2016 Roma. All rights reserved.
//

#import <UIKit/UIKit.h>
#import <CoreData/CoreData.h>
#import <CoreData/NSManagedObject.h>
#import "DownloadManager.h"
#import "DownloadFile.h"
#import "AppDelegate.h"


@interface DownloadManager () <NSURLSessionDelegate, NSFetchedResultsControllerDelegate>

@property (strong, nonatomic) NSURLSession *session;
@end

@implementation DownloadManager



- (instancetype)initSession{
    self = [super init];
    if (self) {
        static dispatch_once_t onceToken;
        dispatch_once(&onceToken, ^{
        
            
            
        NSURLSessionConfiguration *backgroundConfig = [NSURLSessionConfiguration backgroundSessionConfigurationWithIdentifier:[[NSBundle mainBundle] bundleIdentifier]];
       
            static NSInteger maxQuantity = 7;
        backgroundConfig.HTTPMaximumConnectionsPerHost = maxQuantity;
        _maxDownloadQuantity = maxQuantity;
            
        self.session = [NSURLSession sessionWithConfiguration:backgroundConfig
                                                     delegate:self
                                                delegateQueue:nil];
            
        
        self.downloadFiles = [NSMutableArray array];
        });
  }
    
    return self;
}




-(void) addDownloadFileForUrl:(NSString *)urlString
                     withName:(NSString *)titleName
{
    
    DownloadFile *file = [[DownloadFile alloc] initWithTitleName:titleName
                                                      urlString:urlString];
    file.dateStartDownload = [NSDate date];
    
    NSURLRequest *request = [NSURLRequest requestWithURL:file.url];
    
    file.downloadTask = [self.session downloadTaskWithRequest:request];
    
    [self.downloadFiles addObject:file];
    //[self setToCoreData:file];
    
    
}
#pragma mark - CoreData

-(NSManagedObjectContext *) managedObjectContext
{
    
    AppDelegate *appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
    NSManagedObjectContext * context = appDelegate.mainContext;
    
    return context ;
    
}


-(void)setToCoreData:(DownloadFile*)file
{
    NSManagedObjectContext * context = [self managedObjectContext];
    
    NSManagedObject * downloadInfo = [NSEntityDescription insertNewObjectForEntityForName:@"DownloadEntity" inManagedObjectContext:context];
    
    [downloadInfo setValue:file.titleName forKey:@"name"];
    [downloadInfo setValue:file.dateStartDownload forKey:@"dateStartDownload"];
    [downloadInfo setValue:file.dateCompletDownload forKey:@"dateCompletedDownload"];
    [downloadInfo setValue:file.statusCompleted forKey:@"status"];
    
    NSError *error = nil;
    // Save the object to persistent store
    if (![context save:&error]) {
        NSLog(@"Can't Save! %@ %@", error, [error localizedDescription]);
    }
    
}


-(void) downlodInfoForIndex:(NSInteger) index
              progressBlock:(void(^)(CGFloat progress, NSString* write, NSString* total))progressBlock
          downloadTimeBlock:(void(^)(NSString* time))downloadTimeBlock
 isCompletedDowloadindBlock:(void(^)(BOOL isCompleted))isCompletedDownloadingBlock{
    
    DownloadFile * file = [self.downloadFiles objectAtIndex:index];
    if(file)
    {
    file.progressBlock = progressBlock;
    file.downloadTimeBlock = downloadTimeBlock;
    file.isCompletedDownloadingBlock = isCompletedDownloadingBlock;
    }
}

#pragma mark - Action Method
-(void) pauseForIndex:(NSInteger) index{
    DownloadFile* file = [self.downloadFiles objectAtIndex:index];
    [file.downloadTask suspend];
}

-(void) startForIndex:(NSInteger) index{
    DownloadFile* file = [self.downloadFiles objectAtIndex:index];
    [file.downloadTask resume];
}
-(void) removeDownloadFileAtIndex:(NSInteger) index
{
    DownloadFile* file = [self.downloadFiles objectAtIndex:index];
    if (file)
    {
        [file.downloadTask cancel];
        [self.downloadFiles removeObject:file];
    }
}
-(void) cancelForIndex:(NSInteger) index{
    DownloadFile* file = [self.downloadFiles objectAtIndex:index];
    [file.downloadTask cancel];
}

#pragma mark - NSURLSessionDelegate

- (void)URLSession:(NSURLSession *)session
      downloadTask:(NSURLSessionDownloadTask *)downloadTask
      didWriteData:(int64_t)bytesWritten
 totalBytesWritten:(int64_t)totalBytesWritten
totalBytesExpectedToWrite:(int64_t)totalBytesExpectedToWrite {
    

DownloadFile * file = [self getFileForTaskIndetifier:downloadTask.taskIdentifier];
   
    if (file) {
        CGFloat progress = (CGFloat)totalBytesWritten / (CGFloat)totalBytesExpectedToWrite;
        
        dispatch_async(dispatch_get_main_queue(), ^(void) {
            NSLog(@"Progres # %f.2 #, Name:%@ ***",progress, file.titleName );
            NSString* totalMbWritten = [self convertSizeToMB:totalBytesWritten];
            NSString* totalMbExpectedToWrite = [self convertSizeToMB:totalBytesExpectedToWrite];
           if(file.progressBlock)
           {
               file.progressBlock(progress,totalMbWritten,totalMbExpectedToWrite);
           }
                });
    NSString* time = [self timeForDownloadFile:file
                             totalBytesWritten:totalBytesWritten
                     totalBytesExpectedToWrite:totalBytesExpectedToWrite];

          if (file.downloadTimeBlock) {
           dispatch_async(dispatch_get_main_queue(), ^(void) {
                if (file.downloadTimeBlock) {
                    file.downloadTimeBlock((NSString*)time);
                }
             });
         }       
   }

}
- (NSString*)timeForDownloadFile:(DownloadFile *)file
               totalBytesWritten:(int64_t)totalBytesWritten
       totalBytesExpectedToWrite:(int64_t)totalBytesExpectedToWrite {
        
    NSTimeInterval timeInterval = [[NSDate date] timeIntervalSinceDate:file.dateStartDownload];
    
    CGFloat speed = (CGFloat)totalBytesWritten / (CGFloat)timeInterval;
    NSInteger remainedBytes = totalBytesExpectedToWrite - totalBytesWritten;
    NSInteger remainedTimeSecond =  remainedBytes / speed;
    
        NSInteger seconds = remainedTimeSecond % 60;
        NSInteger minutes = (remainedTimeSecond / 60) % 60;
        NSInteger hours = remainedTimeSecond / 3600;
        
        return [NSString stringWithFormat:@"%02ld:%02ld:%02ld",(long)hours, (long)minutes, (long)seconds];
 
}
    

- (void)URLSession:(NSURLSession *)session downloadTask:(NSURLSessionDownloadTask *)downloadTask didFinishDownloadingToURL:(NSURL *)location {

    
    DownloadFile * file = [self getFileForTaskIndetifier:downloadTask.taskIdentifier];

    NSError *error;
    NSFileManager *fileManager = [NSFileManager defaultManager];
    
    NSString *destinationFilename = downloadTask.originalRequest.URL.lastPathComponent;
    NSURL *destinationURL = [[self getDirectoryUrlPath] URLByAppendingPathComponent:destinationFilename];
    
    if ([fileManager fileExistsAtPath:[destinationURL path]]) {
        [fileManager removeItemAtURL:destinationURL error:nil];
    }
    
    BOOL success = [fileManager copyItemAtURL:location
                                        toURL:destinationURL
                                        error:&error];
    
    if (success)
    {
        file.dateCompletDownload = [NSDate date];
        file.statusCompleted = @"Successfully";
        if (file.isCompletedDownloadingBlock) {
            NSLog(@"%@", location);
            dispatch_async(dispatch_get_main_queue(), ^(void){
                file.isCompletedDownloadingBlock(success);

        });
            
            
    }
        else
        {
            file.statusCompleted = @"Fail";
        }
       
        [self setToCoreData:file];
    }
    
}



-(void)URLSession:(NSURLSession *)session task:(NSURLSessionTask *)task didCompleteWithError:(NSError *)error{
    NSString * massage = nil;
    if (error) {
        massage = [NSString stringWithFormat:@"Error: %@", [error localizedDescription]];
    }
    else{
       
        massage = @"Successfully";
    }
   
        [self callCompletionHandlerIfFinished];
}

    - (NSURL *)getDirectoryUrlPath {
        NSArray *paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
        NSString *getDirectory = [paths objectAtIndex:0];
        NSURL *getDirectoryUrl = [NSURL fileURLWithPath:getDirectory];
        return getDirectoryUrl;
    }


- (void)callCompletionHandlerIfFinished
{
    NSLog(@"call completion handler");
    [self.session getTasksWithCompletionHandler:^(NSArray *dataTasks, NSArray *uploadTasks, NSArray *downloadTasks) {
        NSUInteger count = [dataTasks count] + [uploadTasks count] + [downloadTasks count];
        if (count == 0) {
           
            NSLog(@"all tasks ended");
            AppDelegate *appDelegate = (AppDelegate *)[[UIApplication sharedApplication] delegate];
            if (appDelegate.backgroundTransferCompletionHandler == nil) return;
            void (^comletionHandler)() = appDelegate.backgroundTransferCompletionHandler;
            appDelegate.backgroundTransferCompletionHandler = nil;
            comletionHandler();
        }
    }];
}

#pragma mark - Other Method

-(DownloadFile*)getFileForTaskIndetifier:(NSInteger)downloadTask
{
    
    for (int i = 0; i < [self.downloadFiles count]; ++i)
    {
        DownloadFile* download = [self.downloadFiles objectAtIndex:i];
        {
            if (download.downloadTask.taskIdentifier == downloadTask)
            {
                return download;
            }
        }
    }
    return nil;
}

-(NSString*) convertSizeToMB:(int64_t) size
{
    const CGFloat oneMb = 1000000.f;
    CGFloat resultInMB = (CGFloat)size/oneMb;
    return [NSString stringWithFormat:@"%.2f Mb", resultInMB];
}

@end

File: SimpleDM/DownloadManager.swift
//
//  DownloadManager.swift
//  SimpleDM
//
//  Created by Sam on 11/21/16.
//  Copyright © 2016 Roma. All rights reserved.
//

import UIKit
import CoreData


class DownloadManager: NSObject, URLSessionDelegate, URLSessionDownloadDelegate{
    

  
    // MARK: Default initialization
    
     var downloadFiles = NSMutableArray()
     let maxDownloadQuantity = 7
    
      lazy  var downloadSession:URLSession =
        {
            
            struct SessionProperties{
            static let identifier:String! = "myBackgroundSession";
                        }
            
            let sessionConfig = URLSessionConfiguration.background(
                withIdentifier:SessionProperties.identifier)
//            sessionConfig.timeoutIntervalForRequest = 0
//            sessionConfig.timeoutIntervalForResource = 0
    
            return URLSession(configuration: sessionConfig,
                                   delegate: self,
                              delegateQueue: nil)
        }()
  
       // MARK: Creat Methods
    
    
    func addDownloadFileForUrl(urlString: String,
                      withName titleName: String)->Void
    {
        let file = DownloadFile(titleName: titleName, urlString: urlString)
        file.dateStartDownload = NSDate()
        file.downloadTask = self.downloadSession.downloadTask(with: file.url);       downloadFiles.add(file)
    }
    
    func downloadInfoForIndex(index:NSInteger,
                      progressBlocK:@escaping(Float,String)->(),
        isCompletedDownloadingBlock:@escaping(Bool)->(),
                  downloadTimeBlock:@escaping(String)->())
    {
        let file:DownloadFile =
            self.downloadFiles.object(at: index) as! DownloadFile
        
        file.progressBlock = progressBlocK
        file.isCompletedDownloadingBlock = isCompletedDownloadingBlock
        file.downloadTimeBlock = downloadTimeBlock
    }
    
    
    // MARK: CoreDate
    
    func seveIntoCoreDate(file: DownloadFile) -> () {
        let context = (UIApplication.shared.delegate as! AppDelegate).persistentContainer.viewContext
        
        
        let task = DownloadEntity(context:context)
        task.titleName = file.titleName;
        task.dateStart = file.dateStartDownload;
        task.dateFinished = file.dateCompletDownload;
        task.finishStaus = file.statusCompleted;
        
        (UIApplication.shared.delegate as! AppDelegate).saveContext()
    }
   
  // MARK: Action Methods
    
    func startForIndex(index:NSInteger)->Void
    {
       if let file:DownloadFile =
        self.downloadFiles.object(at: index) as? DownloadFile{
        
        file.downloadTask!.resume()
        }
    }
    
    func pauseForIndex(index:NSInteger)->Void
    {
       if let file:DownloadFile =
        self.downloadFiles.object(at: index) as? DownloadFile{
        
        file.downloadTask!.suspend()
        }
    }
    
    func removeDownloadFileAtIndex(index: NSInteger)
    {
        if let file:DownloadFile =
            self.downloadFiles.object(at: index) as? DownloadFile{
         file.downloadTask!.cancel()
         downloadFiles.remove(file)
        }
    }

   // MARK: URLSession
        
        func urlSession(_ session: URLSession,
                     downloadTask: URLSessionDownloadTask,
  didFinishDownloadingTo location: URL)
    {
       
        if let file = getFileForTaskIdentifier(identifier: downloadTask.taskIdentifier)
        {
        var isSeccess = true
        
        let fileManager = FileManager.default
        let destinationURLForFile:URL = getDirectoryUrlPath()
        
        if fileManager.fileExists(atPath: destinationURLForFile.path)
        {
             print(destinationURLForFile.path)
        }
        else{
            do {
                try fileManager.moveItem(at: location,
                                         to: destinationURLForFile)
               print(destinationURLForFile.path)
            }catch{
                isSeccess = false
            }
            
        }
        
        DispatchQueue.global(qos: DispatchQoS.QoSClass.default).async
            {
                
                DispatchQueue.main.async {
                    file.isCompletedDownloadingBlock!(isSeccess)
                }
            
        }
        
        
        file.dateCompletDownload = NSDate();
        file.statusCompleted = isSeccess ? "Successfully" : "Fail"
        seveIntoCoreDate(file:file)
        }
    }

    
    
    
    
    
    
    func urlSession(_ session: URLSession,
                 downloadTask: URLSessionDownloadTask,
    didWriteData bytesWritten: Int64,
            totalBytesWritten: Int64,
    totalBytesExpectedToWrite: Int64){
    
    if let file = getFileForTaskIdentifier(identifier: downloadTask.taskIdentifier)
    {
    let progress = Float(totalBytesWritten)/Float(totalBytesExpectedToWrite)
        
    let written:String = convertSizeToMB(size: totalBytesWritten)
    let total:String = convertSizeToMB(size: totalBytesExpectedToWrite)
    
    let progressStr = String(format: "%@ Mb/%@ Mb",written, total)
        
    let remainedTime:String = timeForDownloadFile(file: file,
                                     totalBytesWritten: totalBytesWritten,
                             totalBytesExpectedToWrite: totalBytesExpectedToWrite)
        
        
        DispatchQueue.global(qos: DispatchQoS.QoSClass.default).async
            {
            
            DispatchQueue.main.async {
                file.progressBlock!(progress,progressStr)
                file.downloadTimeBlock!(remainedTime)
            }
        }
        }
       }
    
    

    func urlSessionDidFinishEvents(forBackgroundURLSession session: URLSession) {
        if let appDelegate = UIApplication.shared.delegate as? AppDelegate {
            if let completionHandler = appDelegate.backgroundSessionCompletionHandler {
                appDelegate.backgroundSessionCompletionHandler = nil
                DispatchQueue.main.async {
                    completionHandler()
                }
            }
        }
    }
    
    
    
    //MARK: Other Methods
    func convertSizeToMB(size: Int64)->String
    {
        let oneMB:Float = 1000000
        let resultInMB = Float(size)/oneMB
        return String(format: "%.2f", resultInMB)
    }
    
    func getFileForTaskIdentifier(identifier: NSInteger)->DownloadFile?
    {
        
        for element in downloadFiles
        {
            if ((element as! DownloadFile).downloadTask?.taskIdentifier == identifier)
            {
                return element as? DownloadFile
            }
        }
      return nil
    }
    
    func getDirectoryUrlPath() -> URL
    {
        let path = NSSearchPathForDirectoriesInDomains(
            FileManager.SearchPathDirectory.documentDirectory,
            FileManager.SearchPathDomainMask.userDomainMask, true)
        
        let documentDirectoryPath:String = path[0]
        let getDirectoryUrl:URL = URL(fileURLWithPath:documentDirectoryPath)
        return getDirectoryUrl
    }
    
    func timeForDownloadFile(file:DownloadFile,
                totalBytesWritten:Int64 ,
        totalBytesExpectedToWrite:Int64)->String    {
        
        let timeInterval = NSDate().timeIntervalSince(file.dateStartDownload! as Date)
        let speed:NSInteger = NSInteger(totalBytesWritten)/Int (timeInterval)
        
        let remainedBytes:NSInteger = totalBytesExpectedToWrite - totalBytesWritten
        let remainedTimeSecond:NSInteger =  remainedBytes / NSInteger(speed)
        
        
        return convertTimeToString(time: remainedTimeSecond)
        
    }
    
    func convertTimeToString(time: NSInteger)->String
    {
        let seconds:NSInteger = time % 60;
        let minutes:NSInteger = (time / 60) % 60;
        let hours:NSInteger = time / 3600;
        return String(format: "%02ld:%02ld:%02ld", hours, minutes,seconds)
    }
  
 
}






